backend/app/models/sample_type.rb

```diff
@@ -0,0 +1,284 @@
+  #
+  # @return the sample types
+  def self.details(id)
+    # GET FEILD TYPES
+    sql = "
+      select *, null as 'allowable_field_types'
+      from field_types ft
+      where ft.parent_class = 'SampleType' and ft.parent_id = #{id}
+      order by ft.name
+    "
+    field_types = FieldType.find_by_sql sql
+
+    # GET SAMPLE_OPTIONS FOR FIELD_TYPE == "SAMPLE"
+    # MAKES <N> CALLS TO THE DB BUT <N> SHOULD NOT BE LARGE
+    field_types.each do |ft|
+      if ft.ftype == "sample"
+        sql = "
+          select aft.id, aft.field_type_id, aft.sample_type_id, st.name
+          from allowable_field_types aft
+          inner join sample_types st on st.id = aft.sample_type_id
+          where aft.field_type_id = #{ft.id}
+          order by st.name
+        "
+        allowable_field_types = AllowableFieldType.find_by_sql sql
+
+        ft = ft.update( {allowable_field_types: allowable_field_types} )
+      end
+    end
+
+    # GET INVENTORY
+    # TODO -- IMPLEMENT THIS WHEN IMPLEMENT INVENTORY
+    inventory = 0
+
+    # GET OBJECT TYPES (CONTAINERS)
+    sql = "
+      select * from object_types where sample_type_id = #{id} order by name
+    "
+    object_types = ObjectType.find_by_sql sql
+
+    return { field_types: field_types, inventory: inventory, object_types: object_types }
+  end
+
+  # Create a sample types.
+  #
+  # @param st [Hash] the sample type
+  #
+  # @option st[:name] [String] the name of the sample type
+  # @option st[:description] [String] the description of the sample type
+  # @option st[:field_types] [Hash] the field_type attributes associated with the sample type
+  #
+  # feild_types = {
+  #   name [String] name
+  #   ftype [String]  ftype
+  #   required [String] required - interpreted as Boolen
+  #   array [String] array - interpreted as Boolean
+  #   choices [String] choices
+  #   allowable_field_types [Array] array of allowable field types (for ftype = "sample")
+  # }
+  #
+  # allowable_field_type = {
+  #   sample_type_id [Int] id of the allowable field type
+  # }
+  #
+  # @return the sample type
+  def self.create(st)
+    name = Input.text(st[:name])
+    description = Input.text(st[:description])
+
+    # CREATE TEH SAMPLE TYPE
+    sample_type_new = SampleType.new
+    sample_type_new.name = name
+    sample_type_new.description = description
+
+    valid = sample_type_new.valid?
+    return false, sample_type_new.errors if !valid
+
+    # SAVE THE SAMPLE TYPE IF IT IS VALID
+    sample_type_new.save
+
+    # SAVE FIELD TYPE IF THERE IS A FIELD NAME
+    if st[:field_types].kind_of?(Array)
+      st[:field_types].each do |ft|
+        fname     = Input.text(ft[:name])
+        ftype     = Input.text(ft[:ftype])
+        frequired = Input.boolean(ft[:required])
+        farray    = Input.boolean(ft[:array])
+        fchoices  = Input.text(ft[:choices])
+
+        if fname != ""
+          field_type_new  = FieldType.new
+
+          field_type_new.parent_id    = sample_type_new.id
```

Comment:
just give the values of the fields as parameters to the call to new

so

```
field_type_new = FieldType.new(
  parent_id: sample_type_new.id,
  fname: name,
  ...
)
```

Reply:
OK.

Reply:
Done.